### PR TITLE
2944 keyboard navigation

### PR DESCRIPTION
**Pull request by LucyChyzhova**
fix: https://github.com/Lundalogik/crm-feature/issues/2944#issuecomment-1231614412
## Review:
- [x] Commits are [atomic](https://seesparkbox.com/foundry/atomic_commits_with_git)
- [x] Commits have the correct *type* for the changes made
- [x] Commits with *breaking changes* are marked as such

### Browsers tested:
(Check any that applies, it's ok to leave boxes unchecked if testing something didn't seem relevant.)

Windows:
- [ ] Chrome
- [ ] Edge
- [ ] Firefox

Linux:
- [ ] Chrome
- [ ] Firefox

macOS:
- [x] Chrome
- [ ] Firefox
- [ ] Safari

Mobile:
- [ ] Chrome on Android
- [ ] iOS

### CODE REVIEW

**Comment by adrianschmidt** [comment]
Copied from Slack conversation, for reference:

> I realised we don't need to bother with portalId, because we have the custom component right there in limel-dock anyway, so we can get a reference straight to it. But then I also realised that limel-dock doesn't know anything about the component, and what it might or might not contain, so we can't go looking for a limel-list that might not even be there. It has to be the consumer that sets the focus I think… :thinking_face:

> Also, we need to make sure that focus is put back on the dock-button when the menu is closed, but that ought to be fairly easy I think.

> Or, no, maybe the consumer shouldn't have to set focus. Maybe limel-dock can find the first focusable element inside the custom component, and just focus that… :thinking_face:

> I think the best way might be to try to use https://github.com/focus-trap/focus-trap. It seems to have some support for shadow-DOM nowadays.

**Comment by vicgeralds** [comment]
> Or, no, maybe the consumer shouldn't have to set focus. Maybe limel-dock can find the first focusable element inside the custom component, and just focus that… :thinking_face:
 
> I think the best way might be to try to use https://github.com/focus-trap/focus-trap. It seems to have some support for shadow-DOM nowadays.

What happened to this discussion? Did you explore this? Why did you decide to let the consumer set the focus?

**Comment by vicgeralds** [comment]
> It seems to have some support for shadow-DOM nowadays.

might still be pitfalls
https://github.com/focus-trap/focus-trap#shadow-dom

> Maybe limel-dock can find the first focusable element inside the custom component

If we can't do this, because the custom component uses shadow dom, I think it's better to avoid shadow dom than forcing the consumer to set focus. It should just be a matter of querying elements having `tabindex` I think.

I haven't quite read up on this topic yet, but I noted that focus behaviour has been "under-specified" until recently
https://blog.whatwg.org/focusing-on-focus

**Comment by vicgeralds** [comment]
We might want to try using [delegatesFocus](https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/delegatesFocus) 

```ts
@Component({
  shadow: { delegatesFocus: true }
})

**Comment by vicgeralds** [comment]
> From the description, it's not entirely clear to me what that does

I agree. But it's quite clear from the name, actually. If it receives focus, it delegates focus to first focusable child. So If I just focus the list, it will work. No need for focus-trap.

image

File: src/components/dock/dock-button/dock-button.tsx [hunk 1]
```diff
@@ -153,4 +172,25 @@ export class DockButton {
...
+        }
+
+        setTimeout(() => {
+            console.log('id', this.portalId);
+            const list: HTMLElement = document.querySelector(
+                `#${this.portalId} limel-menu-surface limel-list`
+            );
+            console.log('LIST', list);
+
+            const firstItem: HTMLElement =
```
Comment (vicgeralds) [inline src/components/dock/dock-button/dock-button.tsx:188]: If we have a reference to the custom component element here, we can just do this:

```typescript
if (element.delegatesFocus) {
    element.focus();
    return;
}
```

**Review by vicgeralds — COMMENTED** [review]
If we assume the component passed in `item.dockMenu.componentName` is decorated using `shadow: { delegatesFocus: true }` we can just give it focus instead of trying to find the first item.

If the item element that should receive focus has `tabindex`, no more work is needed.,

File: src/components/dock/dock-button/dock-button.tsx (hunk 1 above)
Comment (Befkadu1) [reply src/components/dock/dock-button/dock-button.tsx:188]: Yes I confirm that `delegatesFocus` works as we want. I will push the fix shortly.

File: src/components/dock/dock-button/dock-button.tsx [hunk 2]
```diff
@@ -69,6 +78,18 @@ export class DockButton {
         return this.renderButton(this.handleClick);
     }
 
+    @Watch('isOpen')
+    protected openWatcher() {
+        if (!this.isOpen) {
```
Comment (Kiarokh) [inline src/components/dock/dock-button/dock-button.tsx:83]: it works like a charm! Great job!

My only comment is about the interaction design. After the menu is opened, if user press <kbd>esc</kbd>, the focus will not go back to the trigger element. 

Not sure if it's possible to fix, but it'd be great if user could close the menu using the keyboard, and continue their navigation pattern. This could however be fixed in another PR too.

File: src/components/dock/dock-button/dock-button.tsx (hunk 1 above)
Comment (adrianschmidt) [reply src/components/dock/dock-button/dock-button.tsx]: I added `shadow: { delegatesFocus: true }` to the menu component in the example.

**Review by adrianschmidt — APPROVED** [review]
